proposals/infra/INF-NNNN-availability-diagnostics.md

```diff
@@ -0,0 +1,272 @@
+<!-- {% raw %} -->
+
+# Strict Availability Diagnostics
+
+* Proposal: [INF-NNNN](INF-NNNN-availability-diagnostics.md)
+* Author(s): [Chris Bieneman](https://github.com/llvm-beanz)
+* Sponsor: [Chris Bieneman](https://github.com/llvm-beanz)
+* Status: **Under Consideration**
+* Planned Version: 20YY (Clang-only)
+
+## Introduction
+
+Strict availability diagnostics introduces two new modes for diagnosing the use
+of unavailable shader APIs. Unavailable shader APIs are APIs that are exposed in
+HLSL code but are not available in the target shader stage or shader model
+version.
+
+## Motivation
+
+Today the enforcement of API availability is the responsibility of bytecode
+validators, runtimes and drivers operating on optimized shader code. Depending
+contextually on where this enforcement occurs this can result in errors being
+produced either late in compilation when mapping back to source locations is
+difficult or at runtime when the runtime or driver rejects the compiled shader.
+
+Strict availability diagnostics provides more actionable diagnostics for users
+by both having detailed error messages and accurate message locations. It also
+provides diagnostics early and allows implementations to generate meaningful
+diagnostics even without generation of debug information.
+
+> Note: The DXIL validator requires line table information to generate error
+> locations which can be inaccurate. Disabling line table information can have a
+> significant compile-time benefit.
+
+## Proposed solution
+
+### Availability Annotations
+
+In Clang, most shader APIs have header declarations that live in the default
+included `hlsl_intrinsics.h`. All shader API declarations have Clang
+availability attributes denoting version information for when the APIs are
+introduced, deprecated, and removed. This version annotation can be per-stage or
+for all stages. The availability annotations will need to exist on declarations
+regardless of this proposal, as they provide information to AST-based tooling
+such as IDE workflows.
+
+As an example of the annotations in action take the following declaration from
+`hlsl_intrinsics.h`:
+
+```c++
+__attribute__((availability(shadermodel, introduced = 6.0)))
+__attribute__((availability(vulkan, introduced = 1.0)))
+__attribute__((clang_builtin_alias(__builtin_hlsl_wave_active_count_bits)))
+uint WaveActiveCountBits(bool Bit);
+
+// This function is available starting in SM 2.0 for pixel shaders, but SM 6.6
+// for compute, mesh and amplification shaders.
+__attribute__((availability(shadermodel_pixel, introduced = 2.0)))
+__attribute__((availability(shadermodel_compute, introduced = 6.6)))
+__attribute__((availability(shadermodel_mesh, introduced = 6.6)))
+__attribute__((availability(shadermodel_amplification, introduced = 6.6)))
+__attribute__((availability(vulkan, introduced = 1.0)))

```

Comment:
Given the potential for Vulkan execution environment availability to differ from DX shader stages and shader models, might we want to have more specific terms for "shadermodel_*" that could apply separately to DX or Vulkan, invoking the applicable target in the name?  Do these shader models and versions have a 1:1 mapping to something for Vulkan targets?

Reply:
I’m not sure I understand what you’re saying. I’ve added a `vulkan` availability annotation, if we need per-stage vulkan annotations we can add those too.

Reply:
The main thing was that, to me, "shadermodel" sounds more generic than it should be, like it doesn't reference DX somehow to differentiate it from a different target.  In other words, if we have to add something like `vulkan_fragment`, maybe `shadermodel_pixel` should be named something more like `dx_shadermodel_pixel` to make it clear that this shadermodel is a DX-specific thing.

It could just be me not fully understanding the target topology here.  Perhaps shadermodel is something meaningful for Vulkan targets as well?

Reply:
“shadermodel” is already claimed for the DirectX shader model and appears in the target triple, as does “vulkan”. We could consider renaming “shadermodel” to “directx” although I think it is fine as it is. We’ve claimed the term so we don’t need a separate prefix.

Reply:
Actually, come to think of it, I don't think this is the best approach to mark this kind of feature support.  I think we should be identifying feature groupings and tagging operations with that feature group, which would be driven by the details of shader model and target.  For instance, ddx would have something more like `__attribute__((availability(hlsl_derivatives)))`, then `hlsl_derivatives` availability depends on shader model, stage, and potentially even available feature sets (like opting-in to mesh/amp shader derivatives).

What do you think?

Reply:
The availability attribute is designed to operate on the target triple directly, and I think we should keep that pattern. If we want to add further groupings of features, we can do that directly in the hlsl_intrinsics header using preprocessor macros to construct the different attribute formulations. That allows us to keep unnecessary complexity out of the compiler and keeps our use of the attribute consistent with its design and implementation.

Reply:
For the per-stage bit, commented on the other thread.
As for the `shadermodel` vs `directx/dx_shadermodel`, I'm fine with both. I agree with Chris that `shadermodel` is already a claimed DX specific name.
And no, we don't have a 1:1 mapping between shadermodel & vk version, which is why using the `cs_X_Y` for DXC when building SPIR-V saddens me, but that's another story.

So fine for me to have DX use `shadermodel_*` for this.